pkg/destroy/gcp/compute.go

```diff
@@ -0,0 +1,226 @@
+package gcp
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	compute "google.golang.org/api/compute/v1"
+)
+
+type nameAndZone struct {
+	name string
+	zone string
+}
+
+func (n nameAndZone) String() string {
+	return fmt.Sprintf("Name: %s, Zone: %s", n.name, n.zone)
+}
+
+// getZoneName extracts a zone name from a zone URL of the form:
+// https://www.googleapis.com/compute/v1/projects/project-id/zones/us-central1-a
+// where the compute service's basepath is:
+// https://www.googleapis.com/compute/v1/projects/
+// Trimming the URL, leaves a string like: project-id/zones/us-central1-a
+func (o *ClusterUninstaller) getZoneName(zoneURL string) string {
+	path := strings.TrimLeft(zoneURL, o.computeSvc.BasePath)
+	parts := strings.Split(path, "/")
+	if len(parts) >= 3 {
+		return parts[2]
+	}
+	return ""
+}
+
+// getInstanceNameAndZone extracts an instance and zone name from an instance URL in the form:
+// https://www.googleapis.com/compute/v1/projects/project-id/zones/us-central1-a/instances/instance-name
+// After trimming the service's base path, you get:
+// project-id/zones/us-central1-a/instances/instance-name
+func (o *ClusterUninstaller) getInstanceNameAndZone(instanceURL string) (string, string) {
+	path := strings.TrimLeft(instanceURL, o.computeSvc.BasePath)
+	parts := strings.Split(path, "/")
+	if len(parts) >= 5 {
+		return parts[2], parts[4]
+	}
+	return "", ""
+}
+
+func (o *ClusterUninstaller) listClusterInstanceGroups() ([]nameAndZone, error) {
+	return o.listInstanceGroups(o.isClusterResource)
+}
+
+func (o *ClusterUninstaller) listInstanceGroups(filterFunc func(string) bool) ([]nameAndZone, error) {
+	o.Logger.Debugf("Listing instance groups")
+	result := []nameAndZone{}
+	ctx, cancel := contextWithTimeout()
```

Comment:
it would be useful to use the context from the clusteruninstaller.

Reply:
I wanted to have individual timeouts for each API call. I would lose that with a single context.

Reply:
using the clusteruninstaller context as base and adding on top of that..
```
ctx, cancel := context.WithTimeout(o.context(), 1*time.Minute)
```

something like that would be useful to cancel all in progress API calls when clusteruninstaller context expires...

Reply:
ahh, makes sense